stdlib/result.mli

```diff
@@ -0,0 +1,89 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Result values.
+
+    @since 4.08 *)
+
+(** {1:results Results} *)
+
+type ('a, 'e) t = ('a, 'e) result = Ok of 'a | Error of 'e
+(** The type for result values. Either a value [Ok v] or an error [Error e]. *)
+
+val ok : 'a -> ('a, 'e) result
+(** [ok v] is [Ok v]. *)
+
+val error : 'e -> ('a, 'e) result
+(** [error e] is [Error e]. *)
+
+val value : ('a, 'e) result -> default:'a -> 'a
+(** [value r ~default] is [v] if [r] is [Ok v] and [default] otherwise. *)
+
+val get_ok : ('a, 'e) result -> 'a
+(** [get_ok r] is [v] if [r] is [Ok v] and @raise Invalid_argument
+    otherwise. *)
+
+val get_error : ('a, 'e) result -> 'e
+(** [get_error r] is [v] if [r] is [Ok v] and @raise Invalid_argument
+    otherwise. *)
+
+val bind : ('a, 'e) result -> ('a -> ('b, 'e) result) -> ('b, 'e) result
+(** [bind r f] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val join : (('a, 'e) result, 'e) result -> ('a, 'e) result
+(** [join rr] is [r] if [rr] is [Ok r] and [rr] if [rr] is [Error _]. *)
+
+val map : ('a -> 'b) -> ('a, 'e) result -> ('b, 'e) result
+(** [map f r] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val fold : ('a -> 'c) -> ('e -> 'c) -> ('a, 'e) result -> 'c
+(** [fold ok error r] is [ok v] if [r] is [Ok v] and [error e] if [r]
+    is [Error e]. *)
```

Comment:
I found the chosen function names momentarily confusing given that there's an `ok` and an `error` function defined in the module. Perhaps the docs could say `ok_f` and `error_f` or some such.